# A stronger Watson's lemma #

$\newcommand{\e}{\varepsilon}$

### The ordinary Watson's lemma ###
Watson's Lemma as stated and proved in \cite{Bender(1978)} and \cite{copson} runs as follows:

Assume
  $\alpha > -1$, $\beta >0$ and $b>0$.
  If $f(t)$ is a continuous function on $[0,b]$ such that it has asymptotic series expansion
  \begin{equation}
  f(t) \sim t^{\alpha} \sum_{n=0}^\infty a_nt^{\beta n}, \,\,\, t \rightarrow 0^+ \>,
\end{equation}
(and if $b=+\infty$ then $f(t) <k\cdot e^{ct}\, (t\rightarrow +\infty)$ for some positive constants $c$ and $k$), then
\begin{equation}
 \int_0^b f(t)e^{-xt}dt \sim \sum_{n=0}^\infty \dfrac{a_n \Gamma (\alpha + \beta n +1)}{x^{\alpha	+\beta n+1}}, \,\,\, x \rightarrow +\infty
\end{equation}


The proof on page 264 of Bender & Orszag is straightforward.  We will use essentially that same proof to give a stronger result.

### Statement of the stronger version ###


Suppose 
\begin{equation}
I(x) = \int_{t=0}^b f(t) e^{-x t}dt
\end{equation}
exists and is finite for $x > 0$.   We will take $x>0$ and $b >0 $, and allow the case $b = \infty$ which will occasionally require explicit mention.

Suppose now that $\phi_k(t)$ is an asymptotic sequence for $k \ge 0$ as $t \to 0^+$, which is to say that $\phi_{k+1}(t) = o(\phi_k(t))$ as $t \to 0^+$, and suppose moreover that each $\phi_k(t) \ge 0$ for all $t \ge 0$.  Suppose that
\begin{equation}
  f(t) \sim \sum_{n=0}^\infty a_n\phi_n(t), \,\,\, t \rightarrow 0^+ \>.
\end{equation}
Suppose also that $\psi_k(x) := \int_{t=0}^c \phi_k(t)\exp(-xt)dt$ (here $c$ might be $b$, or $\infty$, or any convenient nonzero upper limit) is an asymptotic sequence for $k\ge 0$ as $x \to \infty$.  Finally, suppose that the $\psi_k(x)$ decay to zero more slowly than $\exp( -\alpha x )$ for any $\alpha > 0$.  That is, 
\begin{equation}
e^{\alpha x} \psi_k(x) \to \infty
\end{equation}
as $x \to \infty$, for any integer $k\ge 0$ and for any real $\alpha > 0$.

Then
\begin{equation}
 \int_0^b f(t)e^{-xt}dt \sim \sum_{n=0}^\infty a_n \psi_n(x), \,\,\, x \rightarrow +\infty\>.
\end{equation}

### Proof ###

Let all quantities be as denoted above. Then suppose $\e >0$ and consider $I(x,\e) := \int_{t=0}^\e f(t) \exp(-xt)dt$.  We have $I(x)-I(x,\e) = \int_{t=\e}^b f(t)\exp(-xt)dt = \exp(-x\e)\int_{\tau=0}^{b-\e} f(\tau+\e)\exp(-x\tau)d\tau$ which is exponentially smaller than $I(x)$ as $x \to \infty$, for any $\e > 0$.  By hypothesis, this error is also exponentially smaller than any $\psi_k(x)$.

Now choose $\e >0$ (also smaller than $c$) such that the error $R(t,N)$ of the first $N+1$ terms of the asymptotic series 
\begin{equation}
R(t,N) := f(t) - \sum_{k=0}^N  a_k \phi_k(t)
\end{equation}
satisfies $|R(t)| \le K \phi_{N+1}(t)$ for some positive $K$. Then
\begin{equation}
\left| I(x,\e) - \sum_{k=0}^N a_k \int_{t=0}^\e \phi_k(t) e^{-xt}dt \right|   \le K \int_{t=0}^\e \phi_{N+1}(t) e^{-xt} dt
\end{equation}
where we have already used the nonnegativity of $\phi_k(t)$.  We may use it further to observe that the integral on the right-hand side is less than $\int_{t=0}^c \phi_{N+1}(t)\exp(-xt)dt = \psi_{N+1}(x)$, and so
\begin{equation}
\left| I(x,\e) - \sum_{k=0}^N a_k \int_{t=0}^\e \phi_k(t) e^{-xt}dt \right|   \le K\psi_{N+1}(x)\>.
\end{equation}
Now we only make an exponentially small change in the left-hand side when we replace the integrals to $t=\e$ with integrals to $t=c$ or to $t=b$. We thus have at last that
\begin{equation}
 I(x) - \sum_{k=0}^N a_k \psi_k(x) \ll \psi_{N+1}(x)\>.
\end{equation}
Since $N$ was arbitrary, we have proved the strong Watson lemma.
<div style="text-align: right"> $\natural$ </div>

The following code is simple but perhaps not optimal.  The routine "int" will refuse to compute a long answer, however, in the belief that "nobody wants to see that."  It also does a lot of work in classifying the integrand which is potentially unneccesary here.  It was, however, very simple to write, and it works for small examples.  More, it pointed the way to the stronger Watson lemma above.

Copyright (c) 2024 Robert M. Corless

Permission is hereby granted, free of charge, to any person obtaining a copy of this software and associated documentation files (the “Software”), to deal in the Software without restriction, including without limitation the rights to use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the Software, and to permit persons to whom the Software is furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED “AS IS”, WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.

In [1]:
Watson := proc(f::{operator, procedure}, x::name, {N::posint := Order-1}, $) 
  local t, w; 
  w := asympt(f(1/t), t, N+1); 
  w := eval( convert(w, polynom), t=1/t );  
  w := (int(w*exp(-x*t), t = 0 .. infinity) assuming (0 < x)); 
  w := asympt(expand(w), x, N+1) ; 
end proc:

kilobytes used=1800, alloc=5424, time=0.27

In [7]:
Watson( t->sin(t), x, N=6 );

kilobytes used=15884, alloc=14384, time=2.44

$$\frac{\omega}{x^{2}}-\frac{\omega^{3}}{x^{4}}+\frac{\omega^{5}}{x^{6}}$$

In [4]:
Watson( t->sin(sqrt(t)), x, N=3 );

In [5]:
Watson( t->t^sin(t), x, N=2 );

kilobytes used=8843, alloc=18480, time=1.06

For Stirling's original formula:

In [6]:
f := t -> (1/t - 1/(2*sinh(t/2)))/t; 

t -> (1/t-1/2/sinh(1/2*t))/t

In [7]:
SOF := eval( Watson( f, Z ), Z= z+1/2 );

In [8]:
Z := z + 1/2;
lnzf := ln(sqrt(2*Pi)) + Z*ln(Z) - Z - SOF ;

                                 Z := z + 1/2

In [9]:
Digits := 30;
evalf( eval( lnzf, z=10 )/ln( 10! ) );

                                 Digits := 30

In [9]:
F := int( sqrt(t)*exp(-x*t)/(1+t^2), t=0..1 );
# asympt( F, x );

                                 0

$$\int_{0}^{1}\frac{\sqrt{t} {\mathrm e}^{-x t}}{t^{2}+1}d t$$

Yields "Runtime error: Error, (in asympt) unable to compute series" if uncommented.

In [10]:
F4 := convert( Watson( t->sqrt(t)/(1+t^2), x, N=6 ), polynom );

kilobytes used=21526, alloc=18567, time=3.38

$$\frac{\sqrt{\pi} \left(\frac{1}{x}\right)^{\frac{3}{2}}}{2}-\frac{15 \sqrt{\pi} \left(\frac{1}{x}\right)^{\frac{7}{2}}}{8}+\frac{945 \sqrt{\pi} \left(\frac{1}{x}\right)^{\frac{11}{2}}}{32}$$

In [14]:
CodeTools:-Usage( evalf( eval(F, x=13) ) );

memory used=7.69MiB, alloc change=0 bytes, cpu time=78.00ms, real time=93.00ms, gc time=0ns

$$ 0.01852058122$$

In [15]:
CodeTools:-Usage( evalf( eval(F4, x=13)));

memory used=5.09KiB, alloc change=0 bytes, cpu time=0ns, real time=0ns, gc time=0ns

$$ 0.01852687546$$

In [13]:
evalf( eval(F, x=34) )/evalf( eval(F4, x=34));

In [14]:
evalf(34^(-10/2));

In [15]:
dat := [seq( [log10(combinat:-fibonacci(k)), log10(abs(evalf( eval(F, x=combinat:-fibonacci(k)) )/evalf( eval(F4, x=combinat:-fibonacci(k)))-1))],k=4..12)]:
plot( dat, style=point)

kilobytes used=41623, alloc=24691, time=3.73

In [16]:
ah := dat[-1] - dat[-2]:
evalf(ah[2]/ah[1]);

In [17]:
Aye := Int( exp(-x*sin(t)^2), t=0..Pi/2);

In [18]:
value(Aye); # Can Maple evaluate this?

                     1/2 Pi exp(- 1/2 x) BesselI(0, 1/2 x)

Whoa.  Apparently so.

In [19]:
asympt( value(Aye), x );

In [20]:
# IntegrationTools:-Change( Aye, u=sin(t)^2 ) assuming x>0, u>0, u<1;

Yields "Runtime error: Error, (in assuming) when calling 'IntegrationTools:-Change'. Received: 'unable to change variables'" if uncommented

In [21]:
student[changevar]( u=sin(t)^2, Aye, u );

If we call the following with the change of variables u = sin(t)^2 then Maple gets confused, and doesn't know to use the positive square root.  So we have to phrase it correctly.

In [22]:
Student[MultivariateCalculus][ChangeOfVariables]( Aye, [t=arcsin(sqrt(u))] );

In [23]:
Watson( u->1/(2*sqrt(u)*sqrt(1-u)), x, N=6 )

A nice example, from p. 55 of Problems in Perturbation by Nayfeh.  Here the integral is not a known special function in Maple and Maple can only evaluate it numerically.  Numerical evaluation seems to be expensive and takes more than a second to evaluate when $\omega = 100$.  In contrast, the asymptotic formula is accurate and cheap.

In [16]:
Aye := Int( exp(-omega*x^2)*x^(5/2)*ln(1+x), x=1..infinity );

$$\int_{1}^{\infty}{\mathrm e}^{-\omega  \,x^{2}} x^{\frac{5}{2}} \ln \left(1+x \right)d x$$

In [18]:
CodeTools:-Usage( evalf(eval(Aye, omega=130)) );

memory used=17.16MiB, alloc change=0 bytes, cpu time=406.00ms, real time=280.00ms, gc time=187.50ms

$${ 9.359793197\times 10^{-60}}$$

In [25]:
value(Aye);

Nayfeh's artful computation uses Watson's lemma quite elegantly, and gets the first term with extremely little effort.  Here, we must be artful to get Maple to change variables: we use $x = (v+1)^{1/2}$ and not $v = x^2-1$ because Maple loses track of the constraint $1 \le x$ from the integral.

In [26]:
Eye := Student[MultivariateCalculus][ChangeOfVariables]( Aye, [x=(v+1)^(1/2)] );

Once we do that, though, we can get an arbitrary number of terms in the series, essentially effortlessly.

In [27]:
ah := Watson( v->(v+1)^(3/4)*ln(1+(v+1)^(1/2))/2, omega );

In [28]:
latex( exp(-omega)*ah );

{\mathrm e}^{-\omega} \left(\frac{\ln \left(2\right)}{2 \omega}+\frac{\frac{1}{8}+\frac{3 \ln \left(2\right)}{8}}{\omega^{2}}+\frac{\frac{3}{32}-\frac{3 \ln \left(2\right)}{32}}{\omega^{3}}+\frac{\frac{15 \ln \left(2\right)}{128}-\frac{1}{8}}{\omega^{4}}+\frac{-\frac{135 \ln \left(2\right)}{512}+\frac{9}{32}}{\omega^{5}}+O\left(\frac{1}{\omega^{6}}\right)\right)

\begin{equation}
{\mathrm e}^{-\omega} \left(\frac{\ln \left(2\right)}{2 \omega}+\frac{\frac{1}{8}+\frac{3 \ln \left(2\right)}{8}}{\omega^{2}}+\frac{\frac{3}{32}-\frac{3 \ln \left(2\right)}{32}}{\omega^{3}}+\frac{\frac{15 \ln \left(2\right)}{128}-\frac{1}{8}}{\omega^{4}}+\frac{-\frac{135 \ln \left(2\right)}{512}+\frac{9}{32}}{\omega^{5}}\right)
\end{equation}

In [29]:
# plot( Aye/(exp(-omega)*ah), omega=10..100 );

In [30]:
CodeTools:-Usage( evalf( eval( Aye, omega = 100) ) );

memory used=76.25MiB, alloc change=0 bytes, cpu time=672.00ms, real time=535.00ms, gc time=187.50ms

In [31]:
evalf( eval( exp(-omega)*ah, omega=100.0 ) );

In [32]:
F := int( exp(-x*t)*exp(-1/t), t=0..infinity ) assuming x>0;

In [33]:
plot( F, x=0..5 );

In [34]:
latex( map( simplify, asympt( F, x, 3 )) assuming x>0 );

\frac{\sqrt{\pi} {\mathrm e}^{-2 \sqrt{x}}}{x^{\frac{3}{4}}}+\frac{3 \sqrt{\pi} {\mathrm e}^{-2 \sqrt{x}}}{16 x^{\frac{5}{4}}}-\frac{15 \sqrt{\pi} {\mathrm e}^{-2 \sqrt{x}}}{512 x^{\frac{7}{4}}}+\frac{105 \sqrt{\pi} {\mathrm e}^{-2 \sqrt{x}}}{8192 x^{\frac{9}{4}}}-\frac{4725 \sqrt{\pi} {\mathrm e}^{-2 \sqrt{x}}}{524288 x^{\frac{11}{4}}}+O\left(\frac{1}{x^{\frac{13}{4}}}\right)

\begin{equation}
\frac{\sqrt{\pi} {\mathrm e}^{-2 \sqrt{x}}}{x^{\frac{3}{4}}}+\frac{3 \sqrt{\pi} {\mathrm e}^{-2 \sqrt{x}}}{16 x^{\frac{5}{4}}}-\frac{15 \sqrt{\pi} {\mathrm e}^{-2 \sqrt{x}}}{512 x^{\frac{7}{4}}}+\frac{105 \sqrt{\pi} {\mathrm e}^{-2 \sqrt{x}}}{8192 x^{\frac{9}{4}}}-\frac{4725 \sqrt{\pi} {\mathrm e}^{-2 \sqrt{x}}}{524288 x^{\frac{11}{4}}}+O\left(\frac{1}{x^{\frac{13}{4}}}\right)
\end{equation}

In [35]:
#series( exp(-a*ln(t)^2)/t, t) assuming a>0;
# Yields Runtime error: Error, (in assuming) when calling 'series/exp'. Received: 'unable to compute series'

In [36]:
MultiSeries:-series( exp(-a*ln(t)^2)/t, t) assuming a>0;

The problem below has the wrong lower limit&mdash;it works if $1 \le t < \infty$.

In [37]:
int(exp(-x*t)*exp(-a*ln(t)^2)/t, t=0..infinity ) assuming a>0, x>0;

In [38]:
eq := diff(exp(-x*t)*exp(-a*ln(t)^2)/t, t);

In [39]:
simplify( solve(eq,t) );

In [40]:
int( exp(-x*t)/ln(t), t=0..exp(-1) ) assuming x>0;

In [41]:
int( exp(-x*t)/ln(t), t );

In [42]:
hey := solve( diff(exp(-x*t)/ln(t), t ), t );

In [43]:
plot( hey, x=0..4 );

In [44]:
int( exp(-x*t)*ln(t)^2, t=0..infinity ) assuming x>0;

In [45]:
int( exp(-x*t)*ln(ln(t))/ln(t), t=exp(1)..infinity ) assuming x>0;

In [46]:
Watson( t->ln(ln(t+exp(1)))/ln(t+exp(1)), x, N=4 );

In [47]:
Student[MultivariateCalculus][ChangeOfVariables]( Int( exp(-a*ln(t)^2)/t, t=1..infinity), [t=exp(v^(1/2))]);

In [48]:
F := simplify( 1/2*exp(-a*ln(exp(v^(1/2)))^2)/v^(1/2)) assuming v>0;

In [49]:
Watson( v->1/2/sqrt(v), x, N=5); # Obviously works: it's in original Watson form

                              1/2 Pi    (1/x)

In [50]:
int(F,v=0..infinity) assuming a>0;